### DISCUSSION TITLE
Is there a way to package the project using pyinstaller without all of these files?

### DISCUSSION THREAD

**Discussion by zSynctic** [TomSchimansky/CustomTkinter#423] [Q&A] [answered]
I really like customtkinter and the design is so good but I'm trying to make the project public and the program is so simple but there are all of these files tcls and libraries and many other things

**Comment by Akascape** [comment, marked as answer]
@zSynctic I know what the problem is, you want to provide a clean looking executable version of your program but pyinstaller is packing too much files in the package.
I have researched a lot about pyinstaller and I have the possible solutions for you.

First of all use the latest pyinstaller only, there are some few bugs recently fixed in the latest version.

**Solution 1**

Do a fresh installation of a different python version on your system and download only the required modules that your python program needs like customtkinter and pyinstaller. Use pyinstaller in that version to make your executable. 
By this, unnecessary cache and modules will never be included. 

**Solution 2**

If you use `one-folder` mode for your application, then you can delete some folders like the `dist-info`  folders. Those folders just contains the module information which are not that important. I don't recommend to remove other folders like tk or tcl8 because those are interconnected with customtkinter. Most of them don't take much size on the disk. You can distribute that folder in zip format directly.

**Solution 3**

Even after you remove some unnecessary folders, there will still be lots of files in that package. If you want to give a cleaner version then use a `Installer maker` to pack all the files in one executable. It will be a very clean experience for the end user. It will just work like a professional software installation. You can use `Inno Setup` or `Install Forge` to make your installer. 

(Note: You may have to sign the installer otherwise it _may show_ false positives by antiviruses in many cases (specially windows defender). Its because of those unsigned `dll` files made by pyinstaller.  You can warn the user if you don't want to sign it otherwise there is a different solution I can recommend)

**Solution 4**

If you don't want to use installers but still want to give a clean executable, then use the `one-file` mode of Pyinstaller which will create only one executable file. It actually packs the folders in one executable and unpacks them to a temporary folder every time the user opens it. The main issue is that it is a little bit slower than the one-folder mode because it needs to extract the files in a temporary folder. One thing I will recommend everyone is to specify a unique temp folder using the `--runtime-tmpdir` parameter. You can name it just `Temp`. By doing this, it will not create a mess in the default temporary folder of your OS. Windows generally warns the user whenever an application wants to access the main OS temp folder but using a unique temp folder removes that warning message which is a plus point. You can then simply zip that executable version and distribute it.

This is what I like to use every time, there is no need to sign the exe. To my experiments, I never got any false positives (by antiviruses) with the one-file mode.
 
**Conclusion**

That's all the solutions I came up with. There are still some limitations though. 
If you are making a open source public project then I will recommend you to just upload the main python script in your repo and distribute the executable zip through the releases. 

Hope it will help, please mark this as answer if you are satisfied.

**Reply by zSynctic** [reply]
Thanks, but I only have one last question: can I use the one-file mode of pyinstaller when I'm using customtkinter or not?

**Reply by Akascape** [reply]
@zSynctic Yes, it works the same. One file mode is just a compressed version of one folder.

**Reply by zSynctic** [reply]
How can I do that? Because when I read in the wiki of customtkinter it says that only one directory mode works?

**Reply by Akascape** [reply]
@zSynctic I have written a new answer for you related to the one file, hope it will clear the doubts. No need to make one-file if you are happy with the installers.
**Please mark the answer as solution for this QnA.**

**Reply by zSynctic** [reply]
Ok sure thanks for the help

**Comment by Akascape** [comment]
@zSynctic Yes, the one-file mode works with customtkinter, I have built my own customtkinter app in one-file mode only. 

First make sure to **add this function** in you main python code for the assets/images you use.

```
import sys, os
def resource_path(relative_path):
    base_path = getattr(
        sys,
        '_MEIPASS',
        os.path.dirname(os.path.abspath(__file__)))
    return os.path.join(base_path, relative_path)
```

Then use it like this for every asset:

Suppose this is your asset variable

```
asset = "Icon.ico"
```
Then change it to this:
```
asset = resource_path("Icon.ico")
```
We are doing this because in one-file it will have to search the assets within the temp folder only.

**How to make the One-File executable**

I will recommend you to use [Auto-py-to-exe](https://pypi.org/project/auto-py-to-exe/) which is a interface for Pyinstaller. It will make the work easier for you. 

So first choose the py file in the `script location`. Then select `One File` with either `Console Based` or `Window Based` mode. "Window based" will not show any console when we will start our executable. (But sometimes it depends on the type of program)

You can then choose the icon.ico file for your program in the `Icon` section.

The next section is very important, In the `Additional Files` you have to choose all the required files/folders related to your python script like any assets or images including the icon file. For customtkinter applications, you have to manually choose the `customtkinter` module folder/location under the `add folder` option. Other modules will automatically be packed.

Then under the `Settings` section, specify the output directory for your exe.

Thats all you have to do, then just click on `Convert .py to .exe`.

![Screenshot 2022-09-04 205438](https://user-images.githubusercontent.com/89206401/188321124-654cb993-80b3-4c92-8aa4-ca366931c96e.png)

Tip: I will recommend you to also use an advanced option under the `Advanced` tab which is `--runtime tmpdir` and add `Temp` word in it. It will then create temporary folder only in a specific location instead of any OS defined temp folder.

This is the whole process I use to make my applications.

**Reply by zSynctic** [reply]
So do I need to do anything about the libraries I use to include them or do I just use that function and add the assets only for images

**Reply by zSynctic** [reply]
also, I'm using object orientated for my program so do I put the variable out of the class and then when do I change it to the resource_path sorry for the late question I was busy with school and stuff

**Reply by zSynctic** [reply]
I made it work but it requires me to put the ico file along with it in a folder

**Reply by Akascape** [reply]
@zSynctic 
First you have to understand what actually happens in one-file executable of Pyinstaller.

You get one exe file where all the module_folders and files are bundled inside it. When you run your application it extracts the bundled stuff to a separate temporary folder named as "_MEI....". This extraction normally happens in the OS Temp folder if you don't use a specific '--temp_dir'. It also clears that temp folder when the program is closed.

Now as you have packed your external assets in that exe which means the main script will have to **search** for those assets within the new temp folder which will be created after extraction.

So in that case we have to put that **resource** function in front of every external **file path** that we have used in our script.
 
Modules will automatically be imported so we don't need to do anything with those. 

Sometimes we have modules that cannot be packed automatically (like customtkinter) because of some external non-python files associated with them (like .json files).

So that's why we have to add the customtkinter folder separately.

I recently created another application using customtkinter with one-file mode. 

You can refer the source code and practice yourself:
https://github.com/Akascape/Pure-Glitch 

(This is also object oriented)

**Reply by zSynctic** [reply]
Oh ok thanks your source code will help me so much, After adding the ico file and then including it with custom tkinter I still need to put the ico file with the exe or am I doing something wrong?

**Reply by Akascape** [reply]
@zSynctic
You don't have to do anything for the exe icon that you put using "--icon" in pyinstaller. You don't even need to pack it.

But the icon file that your script uses internally needs to be packed and used with the method that I have explained.

You can use different icons like I have used in my case. 

The main tv icon that I used in my application is the icon I put using '--icon' in pyinstaller options.

But the green square icon is used internally which I have packed using that method. 

You can notice the difference.

**Reply by zSynctic** [reply]
The problem now is I tried the method that you used in your program pure-glitch but now when I try to run it I get this issue 
![image](https://user-images.githubusercontent.com/71632495/193427412-21de45d7-d8a7-4122-915c-083a8f8c2606.png)

**Reply by zSynctic** [reply]
The code of the icon is like this: 
![image](https://user-images.githubusercontent.com/71632495/193427532-2e164e2b-c140-4cd7-9ee6-ecb59d69dcb5.png)

**Reply by Akascape** [reply]
@zSynctic 
Your code is correct but you have to add that whole Asset folder in the additional folder options ('--add data') of Pyinstaller just like you did with the customtkinter folder.

**Reply by zSynctic** [reply]
thanks it works now
